File: golang/bm25/bm25_test.go
package bm25

import (
	"math"
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// expectedBm25Cases pins the size of conformance/cases/bm25.json.
const expectedBm25Cases = 4

func TestBm25VectorCount(t *testing.T) {
	var cases []struct {
		Name string `json:"name"`
	}
	conform.Case(t, "bm25.json", &cases)
	if len(cases) != expectedBm25Cases {
		t.Fatalf("bm25.json: got %d cases, want %d", len(cases), expectedBm25Cases)
	}
}

// BM25 ranking is proven against the shared fixture — every case must match the
// Python reference exactly. Mirrors the tokenize EXEMPLAR: load the fixture,
// assert over ALL cases, no leniency, fail loudly on the first mismatch.
func TestBm25Conformance(t *testing.T) {
	var cases []struct {
		Name string `json:"name"`
		Rows []struct {
			EuID string `json:"eu_id"`
			Text string `json:"text"`
		} `json:"rows"`
		Query    string `json:"query"`
		Expected []struct {
			EuID  string  `json:"eu_id"`
			Score float64 `json:"score"`
		} `json:"expected"`
	}
	conform.Case(t, "bm25.json", &cases)

	if len(cases) != expectedBm25Cases {
		t.Fatalf("bm25.json: got %d cases, want %d", len(cases), expectedBm25Cases)
	}

	for _, c := range cases {
		rows := make([]Row, len(c.Rows))
		for i, r := range c.Rows {
			rows[i] = Row{EuID: r.EuID, Text: r.Text}
		}
		got := Rank(rows, c.Query)

		if len(got) != len(c.Expected) {
			t.Fatalf("case %q: got %d results, want %d (%+v)", c.Name, len(got), len(c.Expected), got)
		}
		for i, want := range c.Expected {
			if got[i].EuID != want.EuID {
				t.Fatalf("case %q: result[%d] eu_id = %q, want %q", c.Name, i, got[i].EuID, want.EuID)
			}
			if math.Abs(got[i].Score-want.Score) > 1e-9 {
				t.Fatalf("case %q: result[%d] score = %v, want %v", c.Name, i, got[i].Score, want.Score)
			}
		}
	}
}

File: js/src/bm25/bm25.test.ts
import { describe, it, expect } from "vitest";
import { loadCase } from "../conform/fixtures.js";
import { bm25, type Bm25Row, type Bm25Result } from "./bm25.js";

// BM25 is proven against the shared fixture — every case must match the Python
// reference exactly (scores rounded to 1e-6). Follows the tokenize exemplar:
// load conformance/cases/bm25.json, assert deep equality over ALL cases.
interface Bm25Case {
  name: string;
  rows: Bm25Row[];
  query: string;
  expected: Bm25Result[];
}

/** Vector counts, pinned. A vector silently dropped from the fixture is a
 *  weakened contract that no per-case assertion can see. */
const EXPECTED_COUNTS: Record<string, number> = { cases: 4 };

describe("bm25 conformance", () => {
  const cases = loadCase<Bm25Case[]>("bm25.json");

  it("vector counts are pinned", () => {
    expect({ cases: cases.length }).toEqual(EXPECTED_COUNTS);
  });

  for (const c of cases) {
    it(c.name, () => {
      expect(bm25(c.rows, c.query)).toEqual(c.expected);
    });
  }
});
